client/lib/create-cached-selector/index.js

```diff
@@ -0,0 +1,102 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { isObject, forEach, some, isFunction } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import warn from 'lib/warn';
+
+/*
+ * A map that is Weak with objects but Strong with primitives
+ */
+export class MixedMap {
+	weakMap = new WeakMap();
+	map = new Map();
+
+	constructor( init ) {
+		forEach( init, ( [ key, val ] ) => {
+			this.mapForKey( key ).set( key, val );
+		} );
+	}
+
+	mapForKey = key => ( isObject( key ) ? this.weakMap : this.map );
+
+	clear() {
+		this.weakMap = new WeakMap();
+		this.map.clear();
+		return this;
+	}
+
+	set( k, v ) {
+		this.mapForKey( k ).set( k, v );
+		return this;
+	}
+
+	delete( k ) {
+		this.mapForKey( k ).delete( k );
+		return this;
```

Comment:
do you anticipate chaining these calls like `cache.clear().set().set()` etc…?

I find that it's helpful to _not_ return `this` when we are mutating the underlying object because the code itself kind of indicates based on return value if this is a side-effecting action or a pure one. `moment()` is a good example where it supports chaining _but_ the underlying object mutates throughout the chain. it gives the appearance of being safe from a purity standpoint, but that can lead to mistakes on that assumption.

```js
const then = moment( thenDate )
const nextDay = then.startOf( 'day' ).add( 1, 'day' )
const prevDay = then.startOf( 'day' ).subtract( 1, 'day' )
```

^^^ looks fine until you realize that `prevDay` is actually the start of `thenDate`

Reply:
> do you anticipate chaining these calls like cache.clear().set().set() etc…?

I don't think returning `this` indicates purity...I don't usually assume purity unless its specifically stated.

The thing I would want most is to adhere to the `WeakMap` specification.  Some of them return this and some don't

- `set` --> returns this
- `get` --> returns the value associated with the key
- `clear` --> returns undefined
- `delete` --> returns true/false depending on if the key was found and removed

Reply:
odd. carry on.